### PR TITLE
Community invite

### PR DESCRIPTION
**Pull request by greatbacon**
This PR adds a first pass to inviting people into communities

NB: This relies on some less specified terms/relations (`friend`, `member`) & the system itself is currently non-consentful (I.e. no chance to reject an invite into a community).

### CODE REVIEW

File: src/communities/community.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import type {Space} from '../spaces/space.js';
+import type {Member} from '../members/member.js';
 
 export interface Community {
 	community_id?: number;
 	name: string;
 	spaces: Space[];
+	members: Member[];
```
Comment (ryanatkn) [inline src/communities/community.ts:8]: Nice I like this, clear place to compose a `Persona` in each `Member`.

File: src/db/Database.ts [hunk 2]
```diff
@@ -37,10 +38,11 @@ export class Database {
 				let communities: Community[] = unwrap(
 					await this.repos.communities.filterByAccount(account.account_id!),
 				);
+				let friends: Member[] = unwrap(await this.repos.members.getAll());
```
Comment (ryanatkn) [inline src/db/Database.ts:41]: Nothing to change here, just thinking out loud:

If a `Member` encapsulates a `Persona`'s relationship to one `Community`, I'm wondering what the eventual abstraction for friends will be. If we reuse the `Community` concept like we briefly talked about, so each user gets their own personal `Community` that contextually has its own `Members` to model other people, this seems right on.

File: src/db/Database.ts [hunk 3]
```diff
@@ -74,6 +76,23 @@ export class Database {
 				return {ok: false, type: 'noAccountFound', reason: `No account found with name: ${name}`};
 			},
 		},
+		members: {
+			getAll: async (): Promise<Result<{value: Member[]}, {reason: string}>> => {
```
Comment (ryanatkn) [inline src/db/Database.ts:80]: Don't need to change this now, I'm guessing this should take the community id as an arg to scope it.

File: src/db/Database.ts [hunk 4]
```diff
@@ -74,6 +76,23 @@ export class Database {
...
+				const data = await this.sql<Member[]>`
+				select account_id, name from accounts
+				`;
+				return {ok: true, value: data};
+			},
+			create: async (account_id: number, community_id: string): Promise<Result<{value: any}>> => {
+				const data = await this.sql`
```
Comment (ryanatkn) [inline src/db/Database.ts:87]: I think `{value: Member}` and `this.sql<Member>` are the correct types?

File: src/lib/CommunityNav.svelte [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
 	import type {Community} from '../communities/community.js';
 	import Modal from '$lib/Modal.svelte';
+	import type {Member} from 'src/members/';
 
+	export let friends: Member[];
 	export let communities: Community[];
 	export let selectedCommunity: Community;
 	export let selectCommunity: (community: Community) => void;
 
+	let invitableFriends = friends.filter(
```
Comment (ryanatkn) [inline src/lib/CommunityNav.svelte:11]: To make this reactive, change to `$: invitableFriends`.

File: src/routes/index.svelte [hunk 6]
```diff
@@ -4,12 +4,15 @@
...
 	import type {ClientAccount} from 'src/session/clientSession.js';
 	import type {Community} from 'src/communities/community.js';
+	import type {Member} from 'src/members/member.js';
 
 	const title = 'felt-server';
 	let account: ClientAccount;
 	$: account = $session?.account;
 	let communities: Community[] = [];
 	$: communities = $session?.communities;
+	let friends: Member[] = [];
```
Comment (ryanatkn) [inline src/routes/index.svelte:14]: Here and 2 lines above, the `= []` initialization should be removed, and ideally we don't need these declarations.

There's one tricky caveat in Svelte with TypeScript right now, where we can't use the non-null assertion operator `!`, which tells TypeScript, "ignore null and undefined values here, just trust me".

Svelte template code cannot use any TypeScript right now, so we can't sprinkle `!` in there, so if you get template errors in this case after removing the declaration line, this should fix it: (but don't add the `!` unless it fixes an error)

`$: friends = $session?.friends!;`

**Review by ryanatkn — APPROVED** [review]
very cool, very social

File: src/lib/CommunityNav.svelte [hunk 7]
```diff
@@ -1,11 +1,17 @@
...
+		(x) => !selectedCommunity.members.some((y) => x.account_id == y.account_id),
```
Comment (ryanatkn) [inline src/lib/CommunityNav.svelte:12]: This is good code, not suggesting to change anything. Given that it's an `O=Nsquared` algorithm that often suggests improvements to data structures or access patterns upstream. Just something to think about when we iterate on the design.

File: src/db/Database.ts (hunk 2 above)
Comment (greatbacon) [reply src/db/Database.ts:41]: I think `Friend` as a concept actually transcends the `Community` model, and instead represents a direct relationship between two `Persona`s

File: src/db/Database.ts (hunk 3 above)
Comment (greatbacon) [reply src/db/Database.ts:80]: I think I have a comment on whatever calls this method, but I'll go ahead and put one here too.

File: src/db/Database.ts (hunk 4 above)
Comment (greatbacon) [reply src/db/Database.ts:87]: Indeed you are correct!

File: src/lib/CommunityNav.svelte (hunk 7 above)
Comment (greatbacon) [reply src/lib/CommunityNav.svelte:12]: Yeah, I'm not sure what the process looks like per se, but my gut says that a lot of session things that are arrays in the front end data store should probably be hashes with the `object_id` as key, or something to that effect.

**Pull request merged**